pallets/liquid-staking-two/src/lib.rs

```diff
@@ -245,47 +255,141 @@ pub mod pallet {
         }
 
-        //todo，record reward on each era, invoked by stake-client
+        /// Invoked by offchain stake client. Declare reward received.
+        ///
+        /// `origin` should be approved account.(not implement yet).
+        ///
+        /// Q: What's the agent?
+        /// Q: Should update exchange_rate?
         #[pallet::weight(10_000)]
         #[transactional]
-        pub fn record_reward(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
-            Ok(().into())
-        }
+        pub fn record_reward(
+            origin: OriginFor<T>,
+            agent: T::AccountId,
+            #[pallet::compact] amount: Balance,
+            block_number: BlockNumber,
+        ) -> DispatchResultWithPostInfo {
+            T::WithdrawOrigin::ensure_origin(origin)?;
+            ensure!(T::Members::contains(&agent), Error::<T>::IllegalAgent);
+            ensure!(
+                !StakingOperationHistory::<T>::contains_key(
+                    Self::current_era(),
+                    StakingOperationType::RecordReward
+                ),
+                Error::<T>::EventRecorded
+            );
+
+            StakingPool::<T>::try_mutate(|m| -> DispatchResult {
+                *m = m.checked_add(amount).ok_or(ArithmeticError::Overflow)?;
+                Ok(())
+            })?;
+
+            let exchange_rate = Rate::checked_from_rational(
+                StakingPool::<T>::get(),
+                T::Currency::free_balance(T::LiquidCurrency::get(), &agent),
+            )
+            .ok_or(Error::<T>::InvalidExchangeRate)?;
 
-        //todo invoked by stake-client, considering insurrance pool
-        #[pallet::weight(10_000)]
-        #[transactional]
-        pub fn record_slash(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+            ExchangeRate::<T>::put(exchange_rate);
+
+            StakingOperationHistory::<T>::insert(
+                Self::current_era(),
+                StakingOperationType::RecordReward,
+                Operation {
+                    status: ResponseStatus::Successed,
+                    amount,
+                    block_number,
+                },
+            );
+            Self::deposit_event(Event::<T>::RewardsRecorded(agent, amount));
             Ok(().into())
         }
 
-        // bond/unbond/rebond/bond_extra may be merge into one
+        /// Invoked by offchain stake client. Declare reward received.
+        ///
+        /// `origin` should be approved account.(not implement yet).
+        ///
+        /// Q: Should update exchange_rate?
         #[pallet::weight(10_000)]
         #[transactional]
-        pub fn record_bond_response(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
-            Ok(().into())
-        }
-
-        #[pallet::weight(10_000)]
-        #[transactional]
-        pub fn record_bond_extra_response(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
-            Ok(().into())
-        }
+        pub fn record_slash(
```

Comment:
This looks similar to `record_reward`. Can we make this function generic to avoid duplicate code.